// revolt.io/esm/src/client/BaseClient.js
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _BaseClient_token;
// deno-lint-ignore-file no-explicit-any
import { deepMerge, EventEmitter, REST } from '../../deps.js';
import { DEFAULT_CLIENT_OPTIONS, Events } from '../util/Constants.js';
export class BaseClient extends EventEmitter {
    constructor(opts = {}) {
        super();
        Object.defineProperty(this, "api", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        _BaseClient_token.set(this, null);
        Object.defineProperty(this, "options", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "bot", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: true
        });
        this.options = deepMerge(DEFAULT_CLIENT_OPTIONS, opts);
        this.api = new REST(this.options.rest);
        this.api.debug = (msg) => this.emit(Events.DEBUG, `[HTTP]: ${msg}`);
    }
    debug(msg) {
        this.emit(Events.DEBUG, `[MAIN]: ${msg}`);
    }
    set token(token) {
        __classPrivateFieldSet(this, _BaseClient_token, token, "f");
        this.api.setToken(token, this.bot);
    }
    get token() {
        return __classPrivateFieldGet(this, _BaseClient_token, "f");
    }
}
_BaseClient_token = new WeakMap();
