// revolt.io/esm/src/client/Client.js
import { BaseClient, WebSocketShard } from './mod.js';
import { ActionManager } from './actions/ActionManager.js';
import { Error } from '../errors/mod.js';
import { ChannelManager, ServerManager, UserManager } from '../managers/mod.js';
export class Client extends BaseClient {
    constructor() {
        super(...arguments);
        Object.defineProperty(this, "ws", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: new WebSocketShard(this)
        });
        Object.defineProperty(this, "actions", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: new ActionManager(this)
        });
        Object.defineProperty(this, "channels", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: new ChannelManager(this)
        });
        Object.defineProperty(this, "servers", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: new ServerManager(this)
        });
        Object.defineProperty(this, "users", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: new UserManager(this)
        });
        Object.defineProperty(this, "user", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: null
        });
        Object.defineProperty(this, "configuration", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "readyAt", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: null
        });
    }
    get readyTimestamp() {
        return this.readyAt?.getTime() ?? null;
    }
    get uptime() {
        return this.readyAt ? Date.now() - this.readyAt.getTime() : null;
    }
    async login(token, type = 'bot') {
        if (!token)
            throw new Error('INVALID_TOKEN');
        this.bot = type.toLowerCase() === 'bot';
        this.token = token ?? null;
        this.debug('Fetch configuration...');
        this.configuration = await this.api.get('/');
        this.debug('Preparing to connect to the gateway...');
        try {
            await this.ws.connect();
        }
        catch (err) {
            await this.destroy();
            throw err;
        }
        this.readyAt = new Date();
    }
    async destroy() {
        this.token = null;
        this.user = null;
        this.readyAt = null;
        await this.ws.destroy();
    }
    isReady() {
        return this.readyAt != null;
    }
}
