// @revoltio/rest/esm/src/REST.js
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var _REST_bot, _REST_token, _REST_queue;
import * as dntShim from "../_dnt.shims.js";
import { CDN } from './CDN.js';
import { DEFAULT_REST_OPTIONS, Queue, stringifyQuery } from './util/mod.js';
import { HTTPError } from './errors/mod.js';
import { deepmerge as merge } from '../deps/deno.land/x/deepmergets@v4.0.3/dist/deno/index.js';
export class REST {
    constructor(options = {}) {
        Object.defineProperty(this, "cdn", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "options", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        _REST_bot.set(this, true);
        _REST_token.set(this, null);
        _REST_queue.set(this, new Queue());
        this.options = merge(DEFAULT_REST_OPTIONS, options);
        this.cdn = new CDN(this.options);
    }
    debug(_msg) { }
    get headers() {
        if (!__classPrivateFieldGet(this, _REST_token, "f")) {
            throw new Error('Expected token to be set for this request, but none was present');
        }
        return {
            [`x-${__classPrivateFieldGet(this, _REST_bot, "f") ? 'bot' : 'session'}-token`]: __classPrivateFieldGet(this, _REST_token, "f"),
        };
    }
    setToken(token, bot = true) {
        __classPrivateFieldSet(this, _REST_token, token, "f");
        __classPrivateFieldSet(this, _REST_bot, bot, "f");
        return this;
    }
    // deno-lint-ignore no-explicit-any
    async request(request) {
        await __classPrivateFieldGet(this, _REST_queue, "f").wait();
        try {
            const controller = new AbortController();
            const timeout = setTimeout(() => controller.abort(), this.options.timeout);
            const res = await dntShim.fetch(request.path, {
                method: request.method,
                headers: this.headers,
                body: request.body ? JSON.stringify(request.body) : void 0,
                signal: controller.signal,
            }).finally(() => clearTimeout(timeout));
            if (res.ok) {
                if (res.headers.get('Content-Type')?.startsWith('application/json')) {
                    return res.json();
                }
                return res.arrayBuffer();
            }
            // TODO: Handle Rate limits
            if (res.status === 429) {
                this.debug(`Hit a 429 while executing a request.
          Method: ${request.method}
          Path: ${request.path}
          Limit: ${this.options.retries}
          Timeout: ${this.options.timeout}ms`);
            }
            if (res.status >= 500 && res.status < 600) {
                if (request.retries === this.options.retries)
                    throw res;
                request.retries++;
                return this.request(request);
            }
            throw res;
        }
        catch (err) {
            if (request.retries === this.options.retries) {
                if (err instanceof dntShim.Response)
                    throw new HTTPError(err, request);
                throw err;
            }
            request.retries++;
            return this.request(request);
        }
        finally {
            __classPrivateFieldGet(this, _REST_queue, "f").next();
        }
    }
    generateRequest(path, opts = {}) {
        const query = opts.query ? `?${stringifyQuery(opts.query)}` : '';
        const options = {
            path: this.options.api + path + query,
            method: opts.method ?? 'GET',
            retries: 0,
            body: opts.body,
            query: opts.query,
        };
        this.debug(`Generating request options for route: ${path} method: ${options.method}`);
        return options;
    }
    get(path, options = {}) {
        return this.request(this.generateRequest(path, { ...options, method: 'GET' }));
    }
    post(path, options = {}) {
        return this.request(this.generateRequest(path, { ...options, method: 'POST' }));
    }
    delete(path, options = {}) {
        return this.request(this.generateRequest(path, { ...options, method: 'DELETE' }));
    }
    put(path, options = {}) {
        return this.request(this.generateRequest(path, { ...options, method: 'PUT' }));
    }
    patch(path, options = {}) {
        return this.request(this.generateRequest(path, { ...options, method: 'PATCH' }));
    }
}
_REST_bot = new WeakMap(), _REST_token = new WeakMap(), _REST_queue = new WeakMap();
