// revolt.io/esm/src/util/Constants.js
import { UUID } from './UUID.js';
export var Events;
(function (Events) {
    Events["CHANNEL_CREATE"] = "channelCreate";
    Events["CHANNEL_DELETE"] = "channelDelete";
    Events["CHANNEL_UPDATE"] = "channelUpdate";
    Events["DEBUG"] = "debug";
    Events["ERROR"] = "error";
    Events["GROUP_JOIN"] = "groupJoin";
    Events["GROUP_LEAVE"] = "groupLeave";
    Events["MESSAGE"] = "message";
    Events["MESSAGE_DELETE"] = "messageDelete";
    Events["MESSAGE_DELETE_BULK"] = "messageDeleteBulk";
    Events["MESSAGE_UPDATE"] = "messageUpdate";
    Events["RAW"] = "raw";
    Events["READY"] = "ready";
    Events["ROLE_CREATE"] = "roleCreate";
    Events["ROLE_DELETE"] = "roleDelete";
    Events["ROLE_UPDATE"] = "roleUpdate";
    Events["SERVER_CREATE"] = "serverCreate";
    Events["SERVER_DELETE"] = "serverDelete";
    Events["SERVER_MEMBER_JOIN"] = "serverMemberJoin";
    Events["SERVER_MEMBER_LEAVE"] = "serverMemberLeave";
    Events["SERVER_MEMBER_UPDATE"] = "serverMemberUpdate";
    Events["SERVER_UPDATE"] = "serverUpdate";
    Events["TYPING_START"] = "typingStart";
    Events["TYPING_STOP"] = "typingStop";
    Events["USER_UPDATE"] = "userUpdate";
})(Events || (Events = {}));
export var WSEvents;
(function (WSEvents) {
    WSEvents["AUTHENTICATE"] = "Authenticate";
    WSEvents["AUTHENTICATED"] = "Authenticated";
    WSEvents["BEGIN_TYPING"] = "BeginTyping";
    WSEvents["BULK"] = "Bulk";
    WSEvents["CHANNEL_ACK"] = "ChannelAck";
    WSEvents["CHANNEL_CREATE"] = "ChannelCreate";
    WSEvents["CHANNEL_DELETE"] = "ChannelDelete";
    WSEvents["CHANNEL_GROUP_JOIN"] = "ChannelGroupJoin";
    WSEvents["CHANNEL_GROUP_LEAVE"] = "ChannelGroupLeave";
    WSEvents["CHANNEL_START_TYPING"] = "ChannelStartTyping";
    WSEvents["CHANNEL_STOP_TYPING"] = "ChannelStopTyping";
    WSEvents["CHANNEL_UPDATE"] = "ChannelUpdate";
    WSEvents["END_TYPING"] = "EndTyping";
    WSEvents["ERROR"] = "Error";
    WSEvents["MESSAGE"] = "Message";
    WSEvents["MESSAGE_BULK_DELETE"] = "BulkMessageDelete";
    WSEvents["MESSAGE_DELETE"] = "MessageDelete";
    WSEvents["MESSAGE_UPDATE"] = "MessageUpdate";
    WSEvents["PING"] = "Ping";
    WSEvents["PONG"] = "Pong";
    WSEvents["READY"] = "Ready";
    WSEvents["SERVER_DELETE"] = "ServerDelete";
    WSEvents["SERVER_MEMBER_JOIN"] = "ServerMemberJoin";
    WSEvents["SERVER_MEMBER_LEAVE"] = "ServerMemberLeave";
    WSEvents["SERVER_MEMBER_UPDATE"] = "ServerMemberUpdate";
    WSEvents["SERVER_ROLE_DELETE"] = "ServerRoleDelete";
    WSEvents["SERVER_ROLE_UPDATE"] = "ServerRoleUpdate";
    WSEvents["SERVER_UPDATE"] = "ServerUpdate";
    WSEvents["USER_RELATIONSHIP"] = "UserRelationship";
    WSEvents["USER_UPDATE"] = "UserUpdate";
})(WSEvents || (WSEvents = {}));
export var ChannelTypes;
(function (ChannelTypes) {
    ChannelTypes["DM"] = "DM";
    ChannelTypes["GROUP"] = "GROUP";
    ChannelTypes["TEXT"] = "TEXT";
    ChannelTypes["VOICE"] = "VOICE";
    ChannelTypes["NOTES"] = "NOTES";
})(ChannelTypes || (ChannelTypes = {}));
export const DEFAULT_CLIENT_OPTIONS = {
    fetchMembers: false,
    rest: {
        api: 'https://api.revolt.chat',
        app: 'https://app.revolt.chat',
        cdn: 'https://autumn.revolt.chat',
        timeout: 15000,
        retries: 3,
    },
    ws: {
        heartbeat: 30000,
        reconnect: true,
    },
};
export const SYSTEM_USER_ID = '0'.repeat(UUID.TIME_LENGTH + UUID.RANDOM_LENGTH);
