// revolt.io/esm/src/util/UUID.js
import { randomBytes } from '../../deps.js';
export class UUID extends null {
    static get PROG() {
        return randomBytes(1).readUInt8() / 0xff;
    }
    static time(now = Date.now()) {
        let mod, result = '';
        for (let i = this.TIME_LENGTH; i > 0; i--) {
            mod = now % this.ENCODING_LENGTH;
            result = this.ENCODING.charAt(mod) + result;
            now = (now - mod) / this.ENCODING_LENGTH;
        }
        return result;
    }
    static hash() {
        let result = '';
        for (let i = this.RANDOM_LENGTH; i > 0; i--) {
            let random = Math.floor(this.PROG * this.ENCODING_LENGTH);
            if (random === this.ENCODING_LENGTH) {
                random = this.ENCODING_LENGTH - 1;
            }
            result = this.ENCODING.charAt(random) + result;
        }
        return result;
    }
    static generate(timestamp = Date.now()) {
        return this.time(timestamp) + this.hash();
    }
    static timestampOf(id) {
        const timestamp = id
            .slice(0, this.TIME_LENGTH)
            .split('')
            .reverse()
            .reduce((carry, char, index) => {
            const encodingIndex = this.ENCODING.indexOf(char);
            if (encodingIndex === -1) {
                throw new Error('invalid character found: ' + char);
            }
            return (carry += encodingIndex * Math.pow(this.ENCODING_LENGTH, index));
        }, 0);
        return new Date(timestamp);
    }
}
Object.defineProperty(UUID, "ENCODING", {
    enumerable: true,
    configurable: true,
    writable: true,
    value: '0123456789ABCDEFGHJKMNPQRSTVWXYZ'
});
Object.defineProperty(UUID, "ENCODING_LENGTH", {
    enumerable: true,
    configurable: true,
    writable: true,
    value: UUID.ENCODING.length
});
Object.defineProperty(UUID, "RANDOM_LENGTH", {
    enumerable: true,
    configurable: true,
    writable: true,
    value: 16
});
Object.defineProperty(UUID, "TIME_LENGTH", {
    enumerable: true,
    configurable: true,
    writable: true,
    value: 10
});
Object.defineProperty(UUID, "TIME_MAX", {
    enumerable: true,
    configurable: true,
    writable: true,
    value: Math.pow(2, 48) - 1
});
