// revolt.io/esm/src/structures/DMChannel.js
import { Channel } from './mod.js';
import { MessageManager, } from '../lib.js';
import { ChannelTypes, DEFAULT_PERMISSION_DM, } from '../util/mod.js';
export class DMChannel extends Channel {
    constructor(client, data) {
        super(client);
        Object.defineProperty(this, "type", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: ChannelTypes.DM
        });
        Object.defineProperty(this, "active", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "permissions", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: DEFAULT_PERMISSION_DM
        });
        Object.defineProperty(this, "messages", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: new MessageManager(this)
        });
        Object.defineProperty(this, "lastMessageId", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: null
        });
        this._patch(data);
    }
    _patch(data) {
        super._patch(data);
        if (typeof data.active === 'boolean')
            this.active = data.active;
        if (data.last_message_id)
            this.lastMessageId = data.last_message_id;
        return this;
    }
    get lastMessage() {
        if (!this.lastMessageId)
            return null;
        return this.messages.cache.get(this.lastMessageId) ?? null;
    }
    bulkDelete(messages) {
        return this.messages.bulkDelete(messages);
    }
    send(options) {
        return this.messages.send(options);
    }
}
