// revolt.io/esm/src/structures/Invite.js
import { Base } from './Base.js';
export class Invite extends Base {
    constructor(client, data) {
        super(client);
        Object.defineProperty(this, "serverId", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: null
        });
        Object.defineProperty(this, "inviterId", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "channelId", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this._patch(data);
    }
    _patch(data) {
        super._patch(data);
        if (data.channel)
            this.channelId = data.channel;
        if (data.creator)
            this.inviterId = data.channel;
        return this;
    }
    get server() {
        return this.client.servers.cache.get(this.serverId) ?? null;
    }
    get channel() {
        return this.client.channels.cache.get(this.channelId) ?? null;
    }
    get inviter() {
        return this.client.users.cache.get(this.inviterId) ?? null;
    }
    toString() {
        return this.client.api.cdn.invite(this.id);
    }
}
