// revolt.io/esm/src/structures/MessageEmbed.js
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _MessageEmbed_url, _MessageEmbed_title, _MessageEmbed_description, _MessageEmbed_icon_url, _MessageEmbed_color, _MessageEmbed_media;
export class MessageEmbed {
    constructor() {
        _MessageEmbed_url.set(this, void 0);
        _MessageEmbed_title.set(this, void 0);
        _MessageEmbed_description.set(this, void 0);
        _MessageEmbed_icon_url.set(this, void 0);
        _MessageEmbed_color.set(this, void 0);
        _MessageEmbed_media.set(this, void 0);
    }
    setTitle(title) {
        __classPrivateFieldSet(this, _MessageEmbed_title, title, "f");
        return this;
    }
    setIcon(iconURL) {
        __classPrivateFieldSet(this, _MessageEmbed_icon_url, iconURL, "f");
        return this;
    }
    setColor(color) {
        __classPrivateFieldSet(this, _MessageEmbed_color, color, "f");
        return this;
    }
    setDescription(description) {
        __classPrivateFieldSet(this, _MessageEmbed_description, description, "f");
        return this;
    }
    setURL(url) {
        __classPrivateFieldSet(this, _MessageEmbed_url, url, "f");
        return this;
    }
    setMedia(media) {
        __classPrivateFieldSet(this, _MessageEmbed_media, media, "f");
        return this;
    }
    toJSON() {
        return {
            title: __classPrivateFieldGet(this, _MessageEmbed_title, "f"),
            description: __classPrivateFieldGet(this, _MessageEmbed_description, "f"),
            url: __classPrivateFieldGet(this, _MessageEmbed_url, "f"),
            icon_url: __classPrivateFieldGet(this, _MessageEmbed_icon_url, "f"),
            colour: __classPrivateFieldGet(this, _MessageEmbed_color, "f"),
            media: __classPrivateFieldGet(this, _MessageEmbed_media, "f"),
        };
    }
}
_MessageEmbed_url = new WeakMap(), _MessageEmbed_title = new WeakMap(), _MessageEmbed_description = new WeakMap(), _MessageEmbed_icon_url = new WeakMap(), _MessageEmbed_color = new WeakMap(), _MessageEmbed_media = new WeakMap();
