// revolt.io/esm/src/structures/Message.js
import { Base, Mentions, } from './mod.js';
import { UUID } from '../util/mod.js';
export class Message extends Base {
    constructor(client, data) {
        super(client);
        Object.defineProperty(this, "type", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: 'TEXT'
        });
        Object.defineProperty(this, "content", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: ''
        });
        Object.defineProperty(this, "channelId", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "authorId", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "embeds", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: []
        });
        Object.defineProperty(this, "attachments", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: []
        });
        Object.defineProperty(this, "mentions", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: new Mentions(this, [])
        });
        Object.defineProperty(this, "editedTimestamp", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: null
        });
        this._patch(data);
    }
    _patch(data) {
        super._patch(data);
        if (Array.isArray(data.embeds)) {
            this.embeds = data.embeds;
        }
        if (Array.isArray(data.attachments)) {
            this.attachments = data.attachments;
        }
        if (Array.isArray(data.mentions)) {
            this.mentions = new Mentions(this, data.mentions);
        }
        if (data.author) {
            this.authorId = data.author;
        }
        if (data.channel) {
            this.channelId = data.channel;
        }
        if (typeof data.content === 'string') {
            this.content = data.content;
        }
        if (data.system) {
            this.type = data.system.type.toUpperCase();
        }
        if (data.edited) {
            this.editedTimestamp = Date.parse(data.edited);
        }
        return this;
    }
    get createdAt() {
        return UUID.timestampOf(this.id);
    }
    get createdTimestamp() {
        return this.createdAt.getTime();
    }
    get editedAt() {
        return this.editedTimestamp ? new Date(this.editedTimestamp) : null;
    }
    get system() {
        return this.type !== 'TEXT';
    }
    get author() {
        return this.client.users.cache.get(this.authorId) ?? null;
    }
    get channel() {
        return this.client.channels.cache.get(this.channelId);
    }
    get serverId() {
        const channel = this.channel;
        return channel.inServer() ? channel.serverId : null;
    }
    get server() {
        return this.client.servers.cache.get(this.serverId) ?? null;
    }
    get member() {
        return this.server?.members.cache.get(this.authorId) ?? null;
    }
    get url() {
        return `https://app.revolt.chat/${this.serverId ? `server/${this.serverId}` : ''}/channel/${this.channelId}/${this.id}`;
    }
    ack() {
        return this.channel.messages.ack(this);
    }
    delete() {
        return this.channel.messages.delete(this);
    }
    reply(content, mention = true) {
        return this.channel.messages.send({
            content,
            replies: [{ id: this.id, mention }],
        });
    }
    edit(options) {
        return this.channel.messages.edit(this, options);
    }
    fetch() {
        return this.channel.messages.fetch(this.id);
    }
    inServer() {
        return this.channel.inServer();
    }
    toString() {
        return this.content;
    }
}
