// revolt.io/esm/src/structures/NotesChannel.js
import { Channel } from './mod.js';
import { MessageManager, } from '../managers/mod.js';
import { ChannelTypes } from '../util/mod.js';
export class NotesChannel extends Channel {
    constructor(client, data) {
        super(client);
        Object.defineProperty(this, "type", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: ChannelTypes.NOTES
        });
        Object.defineProperty(this, "userId", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "lastMessageId", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: null
        });
        Object.defineProperty(this, "messages", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: new MessageManager(this)
        });
        this._patch(data);
    }
    _patch(data) {
        super._patch(data);
        if (data.user) {
            this.userId = data.user;
        }
        return this;
    }
    send(options) {
        return this.messages.send(options);
    }
    bulkDelete(messages) {
        return this.messages.bulkDelete(messages);
    }
    get lastMessage() {
        if (!this.lastMessageId)
            return null;
        return this.messages.cache.get(this.lastMessageId) ?? null;
    }
    get user() {
        return this.client.user;
    }
}
