// revolt.io/esm/src/structures/Role.js
import { Base } from './mod.js';
import { ChannelPermissions, UUID } from '../util/mod.js';
export class Role extends Base {
    constructor(server, data) {
        super(server.client);
        Object.defineProperty(this, "server", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: server
        });
        Object.defineProperty(this, "name", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "color", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: null
        });
        Object.defineProperty(this, "hoist", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: false
        });
        Object.defineProperty(this, "rank", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "overwrite", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this._patch(data);
    }
    _patch(data, clear = []) {
        super._patch(data);
        if (data.name)
            this.name = data.name;
        if (typeof data.hoist === 'boolean')
            this.hoist = data.hoist;
        if (typeof data.rank === 'number')
            this.rank = data.rank;
        if ('colour' in data)
            this.color = data.colour ?? null;
        if (data.permissions) {
            const { a, d } = data.permissions;
            this.overwrite = {
                allow: new ChannelPermissions(a),
                deny: new ChannelPermissions(d),
            };
        }
        for (const field of clear) {
            if (field === 'Colour')
                this.color = null;
        }
        return this;
    }
    get createdAt() {
        return UUID.timestampOf(this.id);
    }
    get createdTimestamp() {
        return this.createdAt.getTime();
    }
    get permissions() {
        return this.overwrite;
    }
    delete() {
        return this.server.roles.delete(this);
    }
    toString() {
        return `<@&${this.id}>`;
    }
}
