// revolt.io/esm/src/structures/Server.js
import { Attachment, Base, Category } from './mod.js';
import { RoleManager, ServerChannelManager, ServerMemberManager, } from '../managers/mod.js';
import { Collection, ServerPermissions, UUID } from '../util/mod.js';
export class Server extends Base {
    constructor(client, data) {
        super(client);
        Object.defineProperty(this, "name", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "description", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: null
        });
        Object.defineProperty(this, "ownerId", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "members", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: new ServerMemberManager(this)
        });
        Object.defineProperty(this, "channels", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: new ServerChannelManager(this)
        });
        Object.defineProperty(this, "roles", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: new RoleManager(this)
        });
        Object.defineProperty(this, "icon", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: null
        });
        Object.defineProperty(this, "banner", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: null
        });
        Object.defineProperty(this, "analytics", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: false
        });
        Object.defineProperty(this, "discoverable", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: false
        });
        Object.defineProperty(this, "nsfw", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: false
        });
        Object.defineProperty(this, "permissions", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "categories", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: new Collection()
        });
        this._patch(data);
    }
    _patch(data, clear = []) {
        super._patch(data);
        if (Array.isArray(data.categories)) {
            this.categories.clear();
            for (const cat of data.categories) {
                const category = new Category(this, cat);
                this.categories.set(category.id, category);
            }
        }
        if (data.icon) {
            this.icon = new Attachment(this.client, data.icon);
        }
        if (data.banner) {
            this.banner = new Attachment(this.client, data.banner);
        }
        if (data.owner) {
            this.ownerId = data.owner;
        }
        if (data.name) {
            this.name = data.name;
        }
        if ('description' in data) {
            this.description = data.description ?? null;
        }
        if (Array.isArray(data.channels)) {
            for (const id of data.channels) {
                const channel = this.client.channels.cache.get(id);
                if (channel?.inServer())
                    this.channels.cache.set(channel.id, channel);
            }
        }
        if (typeof data.roles === 'object') {
            for (const [id, raw] of Object.entries(data.roles)) {
                this.roles._add(Object.assign(raw, { id }));
            }
        }
        if (typeof data.default_permissions === 'number') {
            this.permissions = new ServerPermissions(data.default_permissions)
                .freeze();
        }
        if (typeof data.analytics === 'boolean')
            this.analytics = data.analytics;
        if (typeof data.discoverable === 'boolean') {
            this.discoverable = data.discoverable;
        }
        if (typeof data.nsfw === 'boolean')
            this.nsfw = data.nsfw;
        for (const field of clear) {
            if (field === 'Icon')
                this.icon = null;
            if (field === 'Description')
                this.description = null;
            if (field === 'Banner')
                this.banner = null;
        }
        return this;
    }
    get me() {
        return this.members.cache.get(this.client.user?.id) ?? null;
    }
    get createdAt() {
        return UUID.timestampOf(this.id);
    }
    get createdTimestamp() {
        return this.createdAt.getTime();
    }
    get owner() {
        return this.client.users.cache.get(this.ownerId) ?? null;
    }
    ack() {
        return this.client.servers.ack(this);
    }
    delete() {
        return this.client.servers.delete(this);
    }
    iconURL(options) {
        return this.icon
            ? this.client.api.cdn.icon(this.icon.id, options?.size)
            : null;
    }
    bannerURL(options) {
        return this.banner
            ? this.client.api.cdn.banner(this.banner.id, options?.size)
            : null;
    }
    toString() {
        return this.name;
    }
}
