// revolt.io/esm/src/structures/ServerChannel.js
import { Attachment, Channel, Invite } from './mod.js';
import { ChannelPermissions, Collection } from '../util/mod.js';
export class ServerChannel extends Channel {
    constructor(client, data) {
        super(client);
        Object.defineProperty(this, "name", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "serverId", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "description", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: null
        });
        Object.defineProperty(this, "icon", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: null
        });
        Object.defineProperty(this, "overwrites", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: new Collection()
        });
        Object.defineProperty(this, "nsfw", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: false
        });
        this._patch(data);
    }
    _patch(data, clear = []) {
        super._patch(data);
        if (data.name)
            this.name = data.name;
        if (data.server)
            this.serverId = data.server;
        if ('description' in data)
            this.description = data.description ?? null;
        if (data.icon)
            this.icon = new Attachment(this.client, data.icon);
        if (typeof data.nsfw === 'boolean')
            this.nsfw = data.nsfw;
        if (data.role_permissions) {
            this.overwrites.clear();
            for (const [id, { a, d }] of Object.entries(data.role_permissions)) {
                this.overwrites.set(id, {
                    allow: new ChannelPermissions(a),
                    deny: new ChannelPermissions(d),
                });
            }
        }
        for (const field of clear) {
            if (field === 'Icon')
                this.icon = null;
            if (field === 'Description')
                this.description = null;
        }
        return this;
    }
    async createInvite() {
        const data = await this.client.api.post(`/channels/${this.id}/invites`);
        return new Invite(this.client, data);
    }
    iconURL(options) {
        return this.icon
            ? this.client.api.cdn.icon(this.icon.id, options?.size)
            : null;
    }
    get server() {
        return this.client.servers.cache.get(this.serverId);
    }
    get category() {
        return this.server.categories.find((cat) => cat.children.has(this.id)) ??
            null;
    }
}
