// revolt.io/esm/src/structures/ServerMember.js
import { Base } from './Base.js';
import { Attachment } from './mod.js';
export class ServerMember extends Base {
    constructor(client, data) {
        super(client);
        Object.defineProperty(this, "serverId", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "nickname", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: null
        });
        Object.defineProperty(this, "avatar", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: null
        });
        this._patch(data);
    }
    _patch(data, clear = []) {
        super._patch(data);
        if ('nickname' in data) {
            this.nickname = data.nickname ?? null;
        }
        if (data.avatar) {
            this.avatar = new Attachment(this.client, data.avatar);
        }
        if (data._id) {
            this.serverId = data._id.server;
            this.id = data._id.user;
        }
        for (const field of clear) {
            if (field === 'Avatar')
                this.avatar = null;
            if (field === 'Nickname')
                this.nickname = null;
        }
        return this;
    }
    async setNickname(nickname) {
        await this.server.members.edit(this, { nickname });
        return this;
    }
    ban(reason) {
        return this.server.members.ban(this, reason);
    }
    kick() {
        return this.server.members.kick(this);
    }
    leave() {
        return this.client.servers.delete(this.serverId);
    }
    displayAvatarURL(options) {
        return this.user.displayAvatarURL(options);
    }
    get user() {
        return this.client.users.cache.get(this.id);
    }
    get server() {
        return this.client.servers.cache.get(this.serverId);
    }
    toString() {
        return `<@${this.id}>`;
    }
}
