// revolt.io/esm/src/structures/User.js
import { Attachment, Base, Presence, Status } from './mod.js';
import { Badges, UUID } from '../util/mod.js';
export class User extends Base {
    constructor(client, data) {
        super(client);
        Object.defineProperty(this, "username", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "avatar", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: null
        });
        Object.defineProperty(this, "presence", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: new Presence(this.client)
        });
        Object.defineProperty(this, "badges", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "bot", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: false
        });
        this._patch(data);
    }
    _patch(data, clear = []) {
        super._patch(data);
        if (data.username) {
            this.username = data.username;
        }
        if (data.bot) {
            this.bot = true;
        }
        if (typeof data.badges === 'number') {
            this.badges = new Badges(data.badges).freeze();
        }
        if (data.avatar) {
            this.avatar = new Attachment(this.client, data.avatar);
        }
        if (data.status) {
            this.presence.status = data.status.presence
                ? Status[data.status.presence]
                : Status.Invisible;
            this.presence.text = data.status.text ?? null;
        }
        for (const field of clear) {
            if (field === 'Avatar')
                this.avatar = null;
            if (field === 'StatusText')
                this.presence.text = null;
            if (field === 'StatusPresence')
                this.presence.status = Status.Invisible;
        }
        return this;
    }
    get createdAt() {
        return UUID.timestampOf(this.id);
    }
    get createdTimestamp() {
        return this.createdAt.getTime();
    }
    async block() {
        await this.client.api.put(`/users/${this.id}/block`);
    }
    async unblock() {
        await this.client.api.delete(`/users/${this.id}/block`);
    }
    async createDM() {
        const data = await this.client.api.get(`/users/${this.id}/dm`);
        return this.client.channels._add(data);
    }
    avatarURL(options) {
        return this.avatar
            ? this.client.api.cdn.avatar(this.avatar.id, this.avatar.filename, options?.size)
            : null;
    }
    displayAvatarURL(options) {
        return this.avatarURL(options) ??
            this.client.api.cdn.defaultAvatar(this.id);
    }
    fetch(force = true) {
        return this.client.users.fetch(this, { force });
    }
    toString() {
        return `<@${this.id}>`;
    }
}
