// revolt.io/esm/src/managers/ServerManager.js
import { BaseManager } from './BaseManager.js';
import { TypeError } from '../errors/mod.js';
import { Server } from '../structures/mod.js';
import { UUID } from '../util/mod.js';
export class ServerManager extends BaseManager {
    constructor() {
        super(...arguments);
        Object.defineProperty(this, "holds", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: Server
        });
    }
    _remove(id) {
        const server = this.cache.get(id);
        for (const id of server?.channels.cache.keys() ?? []) {
            this.client.channels._remove(id);
        }
        return super._remove(id);
    }
    async create(name) {
        const { server } = await this.client.api.post('/servers/create', {
            body: {
                name,
                nonce: UUID.generate(),
            },
        });
        return this._add(server);
    }
    async edit(server, options) {
        const id = this.resolveId(server);
        if (!id)
            throw new TypeError('INVALID_TYPE', 'server', 'ServerResolvable');
        await this.client.api.patch(`/servers/${id}`, { body: options });
    }
    async ack(server) {
        const id = this.resolveId(server);
        if (!id)
            throw new TypeError('INVALID_TYPE', 'server', 'ServerResolvable');
        await this.client.api.put(`/servers/${id}/ack`);
    }
    async delete(server) {
        const id = this.resolveId(server);
        if (!id)
            throw new TypeError('INVALID_TYPE', 'server', 'ServerResolvable');
        await this.client.api.delete(`/servers/${id}`);
    }
    async fetch(server, { force = true } = {}) {
        const id = this.resolveId(server);
        if (!id)
            throw new TypeError('INVALID_TYPE', 'server', 'ServerResolvable');
        if (!force) {
            const server = this.cache.get(id);
            if (server)
                return server;
        }
        const data = await this.client.api.get(`/servers/${id}`);
        return this._add(data);
    }
}
