// revolt.io/esm/src/managers/ServerMemberManager.js
import { BaseManager } from './BaseManager.js';
import { TypeError } from '../errors/mod.js';
import { ServerMember, User } from '../structures/mod.js';
import { Collection } from '../util/mod.js';
export class ServerMemberManager extends BaseManager {
    constructor(server) {
        super(server.client);
        Object.defineProperty(this, "server", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: server
        });
        Object.defineProperty(this, "holds", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: ServerMember
        });
    }
    async edit(member, options) {
        const id = this.resolveId(member);
        if (!id) {
            throw new TypeError('INVALID_TYPE', 'member', 'ServerMemberResolvable');
        }
        await this.client.api.patch(`/servers/${this.server.id}/members/${id}`, {
            body: { ...options },
        });
    }
    async ban(member, reason) {
        const id = this.resolveId(member);
        if (!id) {
            throw new TypeError('INVALID_TYPE', 'member', 'ServerMemberResolvable');
        }
        await this.client.api.put(`/servers/${this.server.id}/bans/${id}`, {
            body: { reason },
        });
    }
    async kick(member) {
        const id = this.resolveId(member);
        if (!id) {
            throw new TypeError('INVALID_TYPE', 'member', 'ServerMemberResolvable');
        }
        await this.client.api.delete(`/servers/${this.server.id}/members/${id}`);
    }
    async unban(member) {
        const id = this.resolveId(member);
        if (!id) {
            throw new TypeError('INVALID_TYPE', 'member', 'ServerMemberResolvable');
        }
        await this.client.api.delete(`/servers/${this.server.id}/bans/${id}`);
    }
    async fetch(member) {
        if (typeof member !== 'undefined') {
            const id = this.resolveId(member);
            if (!id) {
                throw new TypeError('INVALID_TYPE', 'member', 'ServerMemberResolvable');
            }
            const data = await this.client.api.get(`/servers/${this.server.id}/members/${id}`);
            return this._add(data);
        }
        const { members } = await this.client.api.get(`/servers/${this.server.id}/members`);
        return members.reduce((coll, cur) => {
            const member = this._add(cur);
            coll.set(member.id, member);
            return coll;
        }, new Collection());
    }
    resolveId(member) {
        if (member == null)
            return null;
        if (member instanceof ServerMember || member instanceof User) {
            return member.id;
        }
        if (typeof member === 'string')
            return member;
        if ('_id' in member)
            return member._id.user;
        return null;
    }
}
